File: wander/shared/src/test/scala/dev/ligature/wander/WanderSuite.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.wander

import dev.ligature.wander.parser.parse
import dev.ligature.wander.lexer.tokenize
import munit.FunSuite

class WanderSuite extends FunSuite {
  val testLexer = true
  val testParser = false
  val testInterpreter = false
  val testOnly: Set[String] = Set() //if set is empty all tests will run
  def runTest(description: String): Boolean =
    testOnly.isEmpty || testOnly.contains(description)

  testData.foreach { testGroup =>
    testGroup.testInstances.foreach { instance =>
      if (testLexer && instance.tokens != null && runTest(instance.description)) {
        test(s"Lexing -- ${testGroup.category} -- ${instance.description}") {
          val tokens = tokenize(instance.script)
          assertEquals(
            tokens,
            Right(instance.tokens),
            s"tokens are not the same for ${instance.description}"
          )
        }
      }

      if (testParser && instance.ast != null && runTest(instance.description)) {
        test(s"Parsing -- ${testGroup.category} -- ${instance.description}") {
          val tokens = tokenize(instance.script)
          tokens match {
            case Left(err) => fail(s"tokenizer failed - ${err.message}")
            case Right(tokens) => {
              val ast = parse(tokens)
              assertEquals(
                ast,
                Right(instance.ast),
                s"AST values are not the same for ${instance.description}"
              )
            }
          }
        }
      }

      if (testInterpreter && runTest(instance.description)) {
        test(
          s"Interpreting -- ${testGroup.category} -- ${instance.description}"
        ) {
          val result = run(instance.script, testGroup.dataset)
          assertEquals(
            result,
            instance.result,
            s"results are not the same for ${instance.description}"
          )
        }
      }
    }
  }
}

File: wander/src/commonTest/kotlin/dev/ligature/wander/interpreter/WanderSuite.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.wander.interpreter

import dev.ligature.wander.library.common
import dev.ligature.wander.library.datasetQueryBindings
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.assertions.withClue
import dev.ligature.wander.run

class WanderSuite: FunSpec() {
  init {
    val testOnly: Set<String> = setOf() // if set is empty all tests will run

    fun runTest(description: String): Boolean =
      testOnly.isEmpty() || testOnly.contains(description)

    testData.forEach { testGroup ->
      testGroup.testInstances.forEach { instance ->
        if (runTest(instance.description)) {
          test(
            "Interpreting -- ${testGroup.category} -- ${instance.description}"
          ) {
            val result = run(instance.script, common())//datasetQueryBindings(testGroup.dataset))
            withClue("results are not the same for ${instance.description}") {
              result shouldBe instance.result
            }
          }
        }
      }
    }
  }
}
